### DISCUSSION TITLE
Optimize constraint question.

### DISCUSSION THREAD

**Discussion by MrDenfish** [kernc/backtesting.py#976] [Q&A]
How would I be able to prevent the following error when I run the optimizer: `ValueError:` Long orders require: SL (0.0109296) < LIMIT (0.01087821) < TP (0.0120336)
`self.buy(size=.99, sl=self.price*(1-(self.sl/100)),tp=self.price*(1+(self.tp/100)))`
   ```
   stats = bt.optimize(shortlen = range(2,9,1),
    longlen = range(10,20,1),
    offset = range(7,48,7),
    sl= range(1,20,1),
    tp= range(1,15,1),
    constraint=lambda p: p.sl < p.tp,
    maximize = 'Equity Final [$]',
    #maximize= optim_func ,
    method='skopt',
    max_tries=200,
    random_state=0,
    return_heatmap=False,
    return_optimization=True)#True when method = skopt)
```

**Comment by kernc** [comment]
The `sl=` and `tp=` levels simply need to be _below_ and _above_ the limit price (in your case current market price, i.e. `self.data.Close[-1]`) adjusted for commissions (in your case 0).
The way you  compute your `self.price` is wrong, or your `sl=` is too tight.
